rllib/core/loss_and_optim/rl_optimizer.py

```diff
@@ -0,0 +1,78 @@
+import abc
+from typing import Mapping, Any, Union
+
+from ray.rllib.core.rl_module import RLModule
+from ray.rllib.core.rl_module.marl_module import MultiAgentRLModule
+from ray.rllib.utils.annotations import (
+    ExperimentalAPI,
+    OverrideToImplementCustomLogic,
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+)
+
+class RLOptimizer(abc.ABC):
+    """Base class for defining a loss function and optimizer for a RLModule.
+    
+    Args:
+        rl_module: The RLModule that will be optimized.
+        config: The configuration for the optimizer.
+
+    Abstract Methods:
+        compute_loss: computing a loss to optimize rl_module over.
+        compute_gradients_and_apply: computing the gradients of the loss,
+            outputted by `compute_loss` and applying them to the rl_module.
+
+    Example:
+    .. code-block:: python
+
+        module = RLModule(...)
+        rl_optim = RLOptimizer(module, ...)
+        sample_batch = ...
+        fwd_out = module.forward_train(sample_batch)
+        loss_dict = rl_optim.compute_loss(fwd_out)
+        infos_and_stats = rl_optim.compute_gradients_and_apply(loss_dict)
+
+        # for checkpointing the RLOptimizer instance
+        rl_optim_state = rl_optim.get_state()
+        rl_optim.set_state(rl_optim_state)
+
+    """
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def __init__(self, module: Union[RLModule, MultiAgentRLModule], config: Mapping[str, Any]):
+        self._module = module
+        self._optim_config = config
+
+    @abc.abstractmethod
+    def compute_optimization_vars(self, fwd_out: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
makes sense.

I think I had made this point earlier ago last week in the design doc but you had said that this would be information leakage and that people can do any necessary computations that they need on fwd_in, inside of `forward_train`.